pkg/envoy/rbac/policy.go

```diff
@@ -1,136 +1,71 @@
-		// This Principal can be defined in terms of one of AND or OR rules.
-		// When AND/OR semantics are not required to define multiple rules corresponding
-		// to this principal, a single Rule in either the AndRules or OrRules will suffice.
-
-		if len(permissionRuleList.AndRules) != 0 && len(permissionRuleList.OrRules) != 0 {
-			return nil, errors.New("Permission rule cannot have both AND & OR rules at the same time")
-		}
-
-		var currentPermission *xds_rbac.Permission
-
-		switch {
-		case len(permissionRuleList.AndRules) != 0:
-			// Combine all the AND rules for this Permission rule with AND semantics
-			var andPermissionRules []*xds_rbac.Permission
-			for _, andPermissionRule := range permissionRuleList.AndRules {
-				// Fill in the destination port permission
-				if andPermissionRule.Attribute == DestinationPort {
-					port, err := strconv.ParseUint(andPermissionRule.Value, 10, 32)
-					if err != nil {
-						return nil, errors.Errorf("Error parsing destination port value %s", andPermissionRule.Value)
-					}
-					portPermission := GetDestinationPortPermission(uint32(port))
-					andPermissionRules = append(andPermissionRules, portPermission)
-				}
-			}
-			currentPermission = andPermissions(andPermissionRules)
+	permissions := make([]*xds_rbac.Permission, 0, len(p.allowedPorts))
+	for _, port := range p.allowedPorts {
+		permissions = append(permissions, GetDestinationPortPermission(port))
+	}
+	if len(permissions) == 0 {
+		// No principals specified for this policy, allow ANY
+		permissions = []*xds_rbac.Permission{getAnyPermission()}
+	}
 
-		case len(permissionRuleList.OrRules) != 0:
-			// Combine all the OR rules for this Permission rule with OR semantics
-			var orPermissionRules []*xds_rbac.Permission
-			for _, orPermissionRule := range permissionRuleList.OrRules {
-				// Fill in the destination port permission
-				if orPermissionRule.Attribute == DestinationPort {
-					port, err := strconv.ParseUint(orPermissionRule.Value, 10, 32)
-					if err != nil {
-						return nil, errors.Errorf("Error parsing destination port value %s", orPermissionRule.Value)
-					}
-					portPermission := GetDestinationPortPermission(uint32(port))
-					orPermissionRules = append(orPermissionRules, portPermission)
-				}
-			}
-			currentPermission = orPermissions(orPermissionRules)
+	policy.Permissions = permissions
 
-		default:
-			// Neither AND/OR rules set, set permission to Any
-			currentPermission = getAnyPermission()
-		}
+	return policy
+}
 
-		finalPermissions = append(finalPermissions, currentPermission)
+// AddPrincipal adds a principal to the list of allowed principals
+func (p *PolicyBuilder) AddPrincipal(principal string) {
+	// We need this extra defense in depth because it is currently possible to configure a wildcard principal
+	// in addition to specific principals. We should instead not allow this, preventing this misconfiguration from
+	// happening through validating webhooks.
```

Comment:
I am unsure what this implies. Is there a problem with the existing code?

Reply:
having both a wildcard and a principal reduces to the wildcard, and specifying extra is unnecessary. 

If you can think of a more clear way to word the comment, please let me know and I'll change it :)